File: content/snippets/python/design_patterns_decorator.py
from dataclasses import dataclass
from typing import Protocol


class Notifier(Protocol):
    def send(self, message: str) -> None: ...


class EmailNotifier:
    def send(self, message: str) -> None:
        print(f"email: {message}")


@dataclass
class SmsDecorator:
    inner: Notifier

    def send(self, message: str) -> None:
        self.inner.send(message)
        print(f"sms: {message}")


SmsDecorator(EmailNotifier()).send("deployment done")

File: content/snippets/go/design_patterns_decorator.go
package main

import "fmt"

type Notifier interface {
	Send(message string)
}

type EmailNotifier struct{}

func (EmailNotifier) Send(message string) {
	fmt.Println("email:", message)
}

type SMSDecorator struct {
	inner Notifier
}

func (d SMSDecorator) Send(message string) {
	d.inner.Send(message)
	fmt.Println("sms:", message)
}

func main() {
	notifier := SMSDecorator{inner: EmailNotifier{}}
	notifier.Send("deployment done")
}
